crossbeam-queue/src/spsc.rs

```diff
@@ -0,0 +1,449 @@
+    #[inline]
+    fn increment(&self, pos: usize) -> usize {
+        if pos < 2 * self.cap - 1 {
+            pos + 1
+        } else {
+            0
+        }
+    }
+
+    /// Returns the distance between two positions.
+    ///
+    /// Positions must be in range `0 .. 2 * cap`.
+    #[inline]
+    fn distance(&self, a: usize, b: usize) -> usize {
+        if a <= b {
+            b - a
+        } else {
+            2 * self.cap - a + b
+        }
+    }
+}
+
+impl<T> Drop for Inner<T> {
+    fn drop(&mut self) {
+        let mut head = self.head.load(Ordering::Relaxed);
+        let tail = self.tail.load(Ordering::Relaxed);
+
+        // Loop over all slots that hold a value and drop them.
+        while head != tail {
+            unsafe {
+                self.slot(head).drop_in_place();
+            }
+            head = self.increment(head);
+        }
+
+        // Finally, deallocate the buffer, but don't run any destructors.
+        unsafe {
+            Vec::from_raw_parts(self.buffer, 0, self.cap);
+        }
+    }
+}
+
+/// Creates a bounded single-producer single-consumer queue with the given capacity.
+///
+/// Returns the producer and the consumer side for the queue.
+///
+/// # Panics
+///
+/// Panics if the capacity is zero.
+///
+/// # Examples
+///
+/// ```
+/// use crossbeam_queue::spsc;
+///
+/// let (p, c) = spsc::<i32>(100);
+/// ```
+pub fn spsc<T>(cap: usize) -> (Producer<T>, Consumer<T>) {
+    assert!(cap > 0, "capacity must be non-zero");
+
+    // Allocate a buffer of length `cap`.
+    let buffer = {
+        let mut v = Vec::<T>::with_capacity(cap);
+        let ptr = v.as_mut_ptr();
+        mem::forget(v);
+        ptr
+    };
+
+    let inner = Arc::new(Inner {
+        head: CachePadded::new(AtomicUsize::new(0)),
+        tail: CachePadded::new(AtomicUsize::new(0)),
+        buffer,
+        cap,
+        _marker: PhantomData,
+    });
+
+    let p = Producer {
+        inner: inner.clone(),
+        head: Cell::new(0),
+        tail: Cell::new(0),
+    };
+
+    let c = Consumer {
+        inner,
+        head: Cell::new(0),
+        tail: Cell::new(0),
+    };
+
+    (p, c)
+}
+
+/// The producer side of a bounded single-producer single-consumer queue.
+///
+/// # Examples
+///
+/// ```
+/// use crossbeam_queue::{spsc, PushError};
+///
+/// let (p, c) = spsc::<i32>(1);
+///
+/// assert_eq!(p.push(10), Ok(()));
+/// assert_eq!(p.push(20), Err(PushError(20)));
+///
+/// assert!(!p.is_empty());
+/// assert!(p.is_full());
+/// ```
+pub struct Producer<T> {
+    /// The inner representation of the queue.
+    inner: Arc<Inner<T>>,
+
+    /// A copy of `inner.head` for quick access.
+    ///
+    /// This value can be stale and sometimes needs to be resynchronized with `inner.head`.
+    head: Cell<usize>,
+
+    /// A copy of `inner.tail` for quick access.
+    ///
+    /// This value is always in sync with `inner.tail`.
+    tail: Cell<usize>,
+}
+
+unsafe impl<T: Send> Send for Producer<T> {}
```

Comment:
Would making just the `Inner` itself `Send` + `Sync` work? That way, you'll need only one impl.

Reply:
I think that would work too, yeah.

Reply:
Leveraging auto-traits is in theory a good things because it helps to catch bugs. Like, if you add `Rc<T>` to `Produce` in the future, the auto-trait approach will catch this (as in, Producer wouldn't be `Send`), while manual usnafe impl will lead to unsound code. For simple cases it doesn't matter though.

Reply:
That's a good point, but it's a double-edged sword. Relying on auto-traits can be risky when all of the following is true:

* Auto-traits are manually implemented for private types.
* Auto-traits are automatically inferred for public types.
* There's a layer of unsafe code between private and public types.

The problem is that the layer of unsafe code can break invariants inferred from auto-traits implemented on private types.

Here's an example. If we relied on `Inner: Send + Sync` and replaced all `Cell<usize>`s with `AtomicUsize`s, then `Sync` would be inferred for `Producer` and `Consumer`, which would be incorrect and could cause data races. That's because the unsafe code in `Producer` and `Consumer` already assumes that they don't implement `Sync`.

So maybe auto-traits are good only when manually implemented on the innermost layer of safe code? But in any case, I think this queue is simple enough that we'd be fine either way.